src/stim/io/measure_record_reader.cc

```diff
@@ -201,8 +225,16 @@ bool MeasureRecordReaderFormatB8::read_bit() {
 }
 
 bool MeasureRecordReaderFormatB8::next_record() {
+    while (!is_end_of_record()) {
+        read_bit();
+    }
     position = 0;
-    return false;
+    int i = fgetc(in);
+    if (i == EOF) {
+        return false;
+    }
+    ungetc(i, in);
```

Comment:
I consider ungetc to be a code smell since it usually indicates that a different and simpler solution is possible.

In this case, we should be able to use `payload` which is the reader's internal one-byte look-ahead buffer:

```c
bool MeasureRecordReaderFormatB8::next_record() {
    while (!is_end_of_record()) read_bit();
    position = 0;
    return payload != EOF;
}
```

I don't think we even need to call `maybe_update_payload` since this is already done in `is_end_of_record` which the while loop calls.

Reply:
`read_bit` raises an exception when it hits EOF in the middle of a record.

Reply:
Basically, in principle I agree but fixing this requires changing the read API to have a "try start record" method that returns a bool indicating whether the data continued or not, and this method has to be called as part of reading the first record. This would be a reasonable change but it is beyond the scope of this particular PR.